### ISSUE TITLE
Snowflake: Add support for CONNECT BY

### ISSUE THREAD

**Issue by karenderer** [sqlfluff/sqlfluff#5863]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

`CONNECT BY` syntax is valid in [Snowflake](https://docs.snowflake.com/en/sql-reference/constructs/connect-by) and is not currently supported in sqlfluff.

### Expected Behaviour

The `CONNECT BY` construct should be parseable.

### Observed Behaviour

When using `CONNECT BY` sqlfluff returns 
> Found unparsable section: 'connect by prior foo = bar'

and 

> WARNING: Parsing errors found and dialect is set to 'snowflake'. Have you configured your dialect correctly?

### How to reproduce

```sql
with tbl as (
    select 'A' as foo, 'B' as bar
    union all
    select 'B' as foo, 'C' as bar
)

select
    *,
    connect_by_root bar as connect_by_root
from tbl
connect by prior foo = bar
```

### Dialect

```
dialect = snowflake
```

### Version

```
$ sqlfluff --version
sqlfluff, version 3.0.6
```

### Configuration

.sqlfluff
```
[sqlfluff]
dialect = snowflake
templater = jinja
sql_file_exts = .sql
rules =
    capitalisation.keywords,
    capitalisation.literals,
    capitalisation.functions,
    capitalisation.types,
    capitalisation.identifiers
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by laurits-inpay** [comment]
sqlfluff currently only allows for `connect by` clauses if they contain the `start with` part as well. That is, however, not mandatory nor useful in every case. The matched grammar rule should we be weakened in that regard to make `start with` optional. I am facing the same issue, so would appreciate a fix.

**Comment by karenderer** [comment]
This appears fixed by https://github.com/sqlfluff/sqlfluff/pull/5996
